### PR TITLE
add bookcorpus

### PR DESCRIPTION
**Pull request by xinyual**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
- [ ] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

cc @dmlc/gluon-nlp-team

### CODE REVIEW

File: scripts/datasets/pretrain_corpus/prepare_bookcorpus.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+
+def cli_main():
+    parser = get_parser()
+    args = parser.parse_args()
+    main(args)
+
+
+if __name__ == "__main__":
+    cli_main()
```
Comment (sxjscience) [inline scripts/datasets/pretrain_corpus/prepare_bookcorpus.py:51]: new line.

File: scripts/pretraining/bert/README.md [hunk 2]
```diff
@@ -1,16 +1,16 @@
...
+2. Phase 1 training with sequence length 128 
 
 ```bash
 python3 run_pretraining.py \
   --model_name google_en_cased_bert_base \
   --optimizer adamw \
   --lr 1e-4 \
   --wd 0.01 \
-  --data data/*.train \
+  --data_dir /BookCorpus/books1/epubtxt \
```
Comment (leezu) [inline scripts/pretraining/bert/README.md:13]:
```suggestion
  --data_dir BookCorpus/books1/epubtxt \
```

The `/` in the beginning seems to be a typo

**Review by leezu — COMMENTED** [review]
Please delete the commented out lines later (before we can merge the PR)

File: scripts/datasets/pretrain_corpus/prepare_bookcorpus.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+"""Prepare the bookcorpus dataset that contains raw text file."""
+
+import os
+import time
+import tarfile
+import argparse
+from gluonnlp.utils.misc import download
+
+_DOWNLOAD_URL\
+    = "https://the-eye.eu/public/AI/pile_preliminary_components/books1.tar.gz"
```
Comment (ZheyuYe) [inline scripts/datasets/pretrain_corpus/prepare_bookcorpus.py:10]: How about put this url into `scripts/datasets/url_checksums` like https://github.com/dmlc/gluon-nlp/blob/master/scripts/datasets/url_checksums/wikipedia.txt

Comment (xinyual) [reply scripts/datasets/pretrain_corpus/prepare_bookcorpus.py]: I have already added it to bookcorpus.txt with hash sha1. I guess now it is more clear.

File: scripts/datasets/pretrain_corpus/prepare_bookcorpus.py [hunk 4]
```diff
@@ -4,11 +4,18 @@
...
 _DOWNLOAD_URL\
     = "https://the-eye.eu/public/AI/pile_preliminary_components/books1.tar.gz"
 
+#_CURR_DIR = os.path.realpath(os.path.dirname(os.path.realpath(__file__)))
+#_URL_FILE_STATS_PATH = os.path.join(_CURR_DIR, '..', 'url_checksums', 'bookcorpus.txt')
+#_URL_FILE_STATS = load_checksum_stats(_URL_FILE_STATS_PATH)
+#print(_URL_FILE_STATS)
+
+#exit()
+
```
Comment (leezu) [inline scripts/datasets/pretrain_corpus/prepare_bookcorpus.py]: Please remove the commented out code which is not needed in the public open source package

File: scripts/pretraining/bert/README.md [hunk 5]
```diff
@@ -19,6 +19,28 @@ python3 run_pretraining.py \
   --max_seq_length 128 \
   --gpus 0,1,2,3,4,5,6,7
 ```
+and can use amp and horovod to accelerate:
+```bash 
+SUBWORD_ALGO=yttm
+SRC=en
+TGT=de
+nohup horovodrun -np 8 -H localhost:8 python run_pretraining.py \
```
Comment (sxjscience) [inline scripts/pretraining/bert/README.md:27]: Is it possible to open a tmux and run the command instead of using nohup?

File: scripts/pretraining/bert/run_pretraining.py [hunk 6]
```diff
@@ -205,6 +207,13 @@ def train(args):
     else:
         get_dataset_fn = get_pretrain_data_npz
 
+    if args.raw:
+        names=os.listdir(args.data_dir)
```
Comment (sxjscience) [inline scripts/pretraining/bert/run_pretraining.py:211]: nit: Add space and it should be `names = os.listdir(args.data_dir)`.

**Comment by sxjscience** [comment]
@xinyual Can I start to review this PR?

File: scripts/datasets/pretrain_corpus/prepare_wikipedia.py [hunk 7]
```diff
@@ -103,6 +109,27 @@ def merge(x):
                             article_lines.append(line)
 
 
+
+class BookscorpusTextFormatting:
```
Comment (sxjscience) [inline scripts/datasets/pretrain_corpus/prepare_wikipedia.py:113]: Will you change the name?

**Review by sxjscience — DISMISSED** [review]
LGTM. @leezu Would you like to review?

File: scripts/processing/segment_sentences.py [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+
+    def segment_articles_into_sentences(self, segmenter):
+        print('Start: Sentence Segmentation')
+        if len(self.articles) is 0:
+            self.load_articles()
+
+        assert len(self.articles) is not 0, 'Please check that input files are present and contain data.'
+
+        # TODO: WIP: multiprocessing (create independent ranges and spawn processes)
+        use_multiprocessing = 'manager'
```
Comment (leezu) [inline scripts/processing/segment_sentences.py:66]: Should we just implement one strategy and delete the code for the others?

It may be even easier to use `multiprocessing.map` compared to the existing implementation. In addition, it will be faster to map over `self.input_files` instead of `self.articles`. The reason is that each `article` in `self.articles` needs to be serialized and sent to the other process. If you map `self.input_files`, only the file name needs to be serialized. This will be much faster. Then, each process can open the file by itself and write to the output file by itself. No need to send the data back to the main process.

File: scripts/processing/segment_sentences.py [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+    def get_sentences_per_shard(self, shard):
+        result = 0
+        for article_id in shard:
+            result += len(self.sentences[article_id])
+
+        return result
+
+
+    def distribute_articles_over_shards(self):
```
Comment (leezu) [inline scripts/processing/segment_sentences.py:155]: Is it necessary for this function to execute on the main process? Why can't each of the `segmenting_num_worker` not directly output data to the shards?

File: scripts/datasets/pretrain_corpus/prepare_bookcorpus.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+                        help="directory for downloaded  files")
+    return parser
+
+
+
+def main(args):
+    url =_URLS['books1']
+    file_hash = _URL_FILE_STATS[url]
+    target_download_location = os.path.join(args.output,
+                                            os.path.basename(url))
```
Comment (leezu) [inline scripts/datasets/pretrain_corpus/prepare_bookcorpus.py:67]: It's better to use a `cache_location` independent from the output location. For example, you can use
```
    parser.add_argument('--cache-path', type=str,
                        default=os.path.join(get_data_home_dir(), 'bookcorpus'),
                        help='The temporary path to download the dataset.')
```

File: scripts/processing/segment_sentences.py [hunk 11]
```diff
@@ -55,77 +100,34 @@ def load_articles(self):
...
+
+        n_processes = self.segmenting_num_worker  # in addition to the main process, total = n_proc+1
+        pool = multiprocessing.Pool(n_processes)
+        rank = 0
+        args=[]
+        for item, name_item in chunks(self.articles, self.output_training_files,
+                                      (len(self.articles) - 1) // (n_processes) + 1,
+                                      (len(self.output_training_files) - 1) // (n_processes) + 1):
+            args.append((item, name_item, rank, segmenter))
+            rank += 1
```
Comment (leezu) [inline scripts/processing/segment_sentences.py:129]: multiprocessing.map supports automatic chunking. Is there any benefit due to manually chunking the articles here?

**Review by leezu — APPROVED** [review]
